validator/client/runner.go

```diff
@@ -77,24 +83,30 @@ func run(ctx context.Context, v iface.Validator) {
 				continue
 			}
 
-			deadline := v.SlotDeadline(slot)
-			slotCtx, cancel := context.WithDeadline(ctx, deadline)
+			deadline := v.SlotDeadline(slot + dutiesDeadline)
+			dutiesCtx, cancel := context.WithDeadline(ctx, deadline)
 
 			var span trace.Span
-			slotCtx, span = prysmTrace.StartSpan(slotCtx, "validator.processSlot")
+			dutiesCtx, span = prysmTrace.StartSpan(dutiesCtx, "validator.processSlot.updateDuties")
 			span.SetAttributes(prysmTrace.Int64Attribute("slot", int64(slot))) // lint:ignore uintcast -- This conversion is OK for tracing.
 
-			log := log.WithField("slot", slot)
-			log.WithField("deadline", deadline).Debug("Set deadline for proposals and attestations")
-
 			// Keep trying to update assignments if they are nil or if we are past an
 			// epoch transition in the beacon node's state.
-			if err := v.UpdateDuties(slotCtx, slot); err != nil {
+			if err := v.UpdateDuties(dutiesCtx, slot); err != nil {
 				handleAssignmentError(err, slot)
-				cancel()
 				span.End()
+				cancel()
 				continue
 			}
+			span.End()
+			cancel()
```

Comment:
wait we want to cancel even when it's successful?

Reply:
Yes, calling cancel() in the happy case is best practice. If you skip it, the timer will eventually clean itself up or get picked up by the GC, but in the meantime, you risk leaking resources

Reply:
You want to call cancel when the work is done, successful or not.